File: common/src/main/java/org/bread_experts_group/breadlib/config/ConfigValue.java
package org.bread_experts_group.breadlib.config;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Objects;
import java.util.function.Supplier;

public record ConfigValue<T>(@NotNull String name, @Nullable Supplier<T> defaultValue, @NotNull String comment) {
	public static <T> Builder<T> builder() {
		return new Builder<>();
	}

	public static class Builder<T> {
		private String name;
		private Supplier<T> value;
		private String comment = "";

		private Builder() {
		}

		public Builder<T> name(String name) {
			this.name = name;
			return this;
		}

		public Builder<T> defaultValue(Supplier<T> value) {
			this.value = value;
			return this;
		}

		public Builder<T> comment(String comment) {
			this.comment = comment;
			return this;
		}

		public ConfigValue<T> build() {
			return new ConfigValue<>(
					Objects.requireNonNull(this.name),
					this.value,
					Objects.requireNonNull(this.comment)
			);
		}
	}
}

File: common/src/main/kotlin/org/bread_experts_group/breadlib/config/ConfigValue.kt
package org.bread_experts_group.breadlib.config

import java.util.function.Supplier

data class ConfigValue<T>(val name: String, val defaultValue: Supplier<T>, val comment: String) {
	companion object {
		fun <T> builder(): Builder<T> = Builder()
	}

	class Builder<T> {
		private lateinit var name: String
		private lateinit var value: Supplier<T>
		private var comment: String = ""

		fun name(name: String): Builder<T> = this.also { this.name = name }
		fun defaultValue(value: Supplier<T>): Builder<T> = this.also { this.value = value }
		fun comment(comment: String): Builder<T> = this.also { this.comment = comment }

		fun build(): ConfigValue<T> = ConfigValue(this.name, this.value, this.comment)
	}
}